Add a scanning strategy to garuntee comprehensive coverage.

rosswsnider opened the pull request.

0xdade commented on natlas-server/app/cyclicprng.py:
```diff
@@ -0,0 +1,89 @@
+from threading import Lock
+import random
+import sympy
```
`sympy` is a new dependency that isn't in the python standard libraries, so this will require an updated requirements.txt

rosswsnider replied:
0xdade did you say you were going to make this change locally? I've refrained from touching/checking-in requirements.txt.

0xdade replied:
You should update your requirements.txt and include it in this PR if you can - that way the merge will be an atomic unit to implement the functionality. (I haven't always been good about this in the past, committing things that I forgot dependencies for)

0xdade commented on natlas-server/app/ipscanmanager.py:
```diff
@@ -0,0 +1,58 @@
+from netaddr import *
+from app.cyclicprng import *
```
I really don't like `import *` if I can help it, because you lose the namespace of the items you're importing, making it more likely to have conflicting names. Can you import explicitly the things you need?

0xdade commented on natlas-server/app/scope.py:
```diff
@@ -1,12 +1,14 @@
 import ipaddress
-
+from netaddr import *
```
See prior comment about `import *`

rosswsnider commented:
Refactored the code a little:
- The import statements are now constrained.
- 0 scope will result in a 404 statement for agents.
- Removed duplicated code between cyclicprng.getNextIP() and cyclicprng.getIP()

rosswsnider commented on natlas-server/app/ipscanmanager.py:
```diff
@@ -0,0 +1,58 @@
+		return binarysearch(self.networks, index)
+
+	def getIP(self, index):
+		def binarysearch(networks, i):
```
Good idea! The code is actually copied because I wanted to test the binary search independent of the PRNG, and then I never removed it because it's a useful method for the IPScanManager! But I never bothered refactoring.
